idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddToStringFixFactories.kt

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.idea.quickfix.fixes
+
+import com.intellij.psi.PsiElement
+import org.jetbrains.kotlin.idea.fir.api.fixes.diagnosticFixFactory
+import org.jetbrains.kotlin.idea.frontend.api.KtAnalysisSession
+import org.jetbrains.kotlin.idea.frontend.api.fir.diagnostics.KtFirDiagnostic
+import org.jetbrains.kotlin.idea.frontend.api.types.KtType
+import org.jetbrains.kotlin.idea.quickfix.AddToStringFix
+import org.jetbrains.kotlin.psi.KtExpression
+
+object AddToStringFixFactories {
+    @OptIn(ExperimentalStdlibApi::class)
+    private fun KtAnalysisSession.getFixes(element: PsiElement?, expectedType: KtType, actualType: KtType): List<AddToStringFix> {
+        if (element !is KtExpression) return emptyList()
+        return buildList {
+            if (expectedType.isString || expectedType.isCharSequence) {
+                add(AddToStringFix(element, false))
+                if (expectedType.isMarkedNullable && actualType.isMarkedNullable) {
+                    add(AddToStringFix(element, true))
+                }
```

Comment:
So, if expected type and actual type are nullable, a user will get two fixes, one with `$0.toString()` and the other with `$0?.toString()`? Should this be:
```kt
add(AddToStringFix(element, expectedType.isMarkedNullable && actualType.isMarkedNullable)
```
?

Reply:
This is simply copying logic from FE1.0 at https://github.com/JetBrains/kotlin/blob/85cbea70bfc7b5814305210b8cf6286bfc6bda2a/idea/src/org/jetbrains/kotlin/idea/quickfix/QuickFixFactoryForTypeMismatchError.kt#L125.

Reply:
This makes sense to me, if you have code like:
```kt
fun foo(s: String?) {}
fun test(i: Int?) {
  foo(i)
}
```
The call can be changed to either `foo(i.toString())` (extension function that returns `"null"`) or `foo(i?.toString())` (simply returns `null`)

Reply:
Oops, comment streams crossed. 🙂  But you know, it _could_ be confusing to the user to offer `i.toString()`. Most of the time users may not want the `"null"` return value there (unless you are trying to render a string), and I don't think it's common knowledge that there is an `Any?.toString()` extension that return `"null"`. So I can see an argument for what @jsjeon is suggesting.

Reply:
Thanks for the example, @punzki . Didn't think about this that way, huh. I'd blame extension functions (`toString`) that allow nullable receiver and silently returns `null` back again. :p